### PR TITLE
FE: Design update for log context tab

### PR DESCRIPTION
**Pull request by Rajat-Dabade**
Design update for log context tab: 

https://github.com/SigNoz/signoz/assets/26198062/2d88bc7d-0912-46da-b766-a0cbaba3001d

## Summary by CodeRabbit

- **New Features**
	- Introduced a new `ContextLogRenderer` component for structured log viewing, including pagination and context-based log rendering.
	- Added styling for the `ContextLogRenderer` and `ContextView` components, enhancing the user interface with customized scrollbars and layout adjustments.
- **Enhancements**
	- Updated the `RawLogView` component to improve log activation logic.
	- Enhanced the `ContextView` component for better usability and maintenance, incorporating the `ContextLogRenderer` for log display.
- **Refactor**
	- Implemented `useContextLogData` for efficient log data management, supporting features like sorting, pagination, and error handling.

### CODE REVIEW

File: frontend/src/container/LogDetailedView/ContextView/ContextLogRenderer.styles.scss [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+        &::-webkit-scrollbar-thumb {
+            background: var(--bg-slate-300);
+        }
+
+        &::-webkit-scrollbar-thumb:hover {
+            background: var(--bg-slate-200);
+        }
+        
+        /* Adjust width to fit content of the third descendant inside virtuoso */
```
Comment (YounixM) [inline frontend/src/container/LogDetailedView/ContextView/ContextLogRenderer.styles.scss:21]: Why?

File: frontend/src/container/LogDetailedView/ContextView/ContextLogRenderer.tsx [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+	return (
+		<div className="context-log-renderer">
+			<ShowButton
+				isLoading={isPreviousLogsFetching}
+				isDisabled={false}
+				order={ORDERBY_FILTERS.ASC}
+				onClick={handlePreviousLogsShowNextLine}
+			/>
+			{isPreviousLogsFetching && <Spinner size="large" height="20rem" />}
```
Comment (YounixM) [inline frontend/src/container/LogDetailedView/ContextView/ContextLogRenderer.tsx:93]: We should be moving towards skeleton designs for loading

File: frontend/src/container/LogDetailedView/ContextView/ContextLogRenderer.tsx [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+			<Virtuoso
+				className="virtuoso-list"
+				initialTopMostItemIndex={0}
+				data={logs}
+				itemContent={getItemContent}
+				style={{ height: `calc(${logs.length} * 32px)` }}
```
Comment (YounixM) [inline frontend/src/container/LogDetailedView/ContextView/ContextLogRenderer.tsx:99]: the height and number of items in virtuoso list should also be dependent on the height of the parent.

File: frontend/src/container/LogDetailedView/ContextView/ContextView.styles.scss [hunk 4]
```diff
@@ -1,3 +1,28 @@
 .log-context-container {
     border: 1px solid var(--bg-slate-400);
-}
\ No newline at end of file
+    height: auto;
+    max-height: 80vh;
```
Comment (YounixM) [inline frontend/src/container/LogDetailedView/ContextView/ContextView.styles.scss:4]: Why ?

File: frontend/src/container/LogDetailedView/ContextView/ContextView.styles.scss [hunk 5]
```diff
@@ -1,3 +1,28 @@
...
+        background: var(--bg-slate-300);
+    }
+    
+    &::-webkit-scrollbar-thumb:hover {
+        background: var(--bg-slate-200);
+    }
+}
+
+.isEditable {
+    height: 73vh; // When edit mode is enabled, reducing the size of log-context-container to avoid unnecessary scroll bars.
```
Comment (YounixM) [inline frontend/src/container/LogDetailedView/ContextView/ContextView.styles.scss:27]: Let's move towards responsive design and avoid fixed heights and widths

File: frontend/src/container/LogDetailedView/ContextView/ContextLogRenderer.tsx [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+		isEdit,
+		query,
+		order: ORDERBY_FILTERS.DESC,
+		page: afterLogPage,
+		setPage: setAfterLogPage,
+	});
+
+	useEffect(() => {
+		setLogs((prev) => [...previousLogs, ...prev]);
+		setFirstLog(logs[0]);
```
Comment (Vikrant2520) [inline frontend/src/container/LogDetailedView/ContextView/ContextLogRenderer.tsx:56]: logs state won't be immediately updated here, is this expected to have firstLog according to the last state here ?

File: frontend/src/container/LogDetailedView/ContextView/ContextLogRenderer.tsx [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, [previousLogs]);
+
+	useEffect(() => {
+		setLogs((prev) => [...prev, ...afterLogs]);
+		setLastLog(logs[logs.length - 1]);
+		// eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (Vikrant2520) [inline frontend/src/container/LogDetailedView/ContextView/ContextLogRenderer.tsx:63]: same here

File: frontend/src/container/LogDetailedView/ContextView/ContextView.styles.scss (hunk 5 above)
Comment (Vikrant2520) [reply frontend/src/container/LogDetailedView/ContextView/ContextView.styles.scss:27]: +1 , never set heights. only play with widths

File: frontend/src/container/LogDetailedView/ContextView/ContextLogRenderer.tsx (hunk 6 above)
Comment (Rajat-Dabade) [reply frontend/src/container/LogDetailedView/ContextView/ContextLogRenderer.tsx:56]: Change it to `previousLogs[0]`

File: frontend/src/container/LogDetailedView/ContextView/ContextLogRenderer.tsx (hunk 7 above)
Comment (Rajat-Dabade) [reply frontend/src/container/LogDetailedView/ContextView/ContextLogRenderer.tsx:63]: Changed this to `afterLogs[afterLogs.length - 1]`

File: frontend/src/container/LogDetailedView/ContextView/ContextLogRenderer.styles.scss (hunk 1 above)
Comment (Rajat-Dabade) [reply frontend/src/container/LogDetailedView/ContextView/ContextLogRenderer.styles.scss:21]: Removed it.

File: frontend/src/container/LogDetailedView/ContextView/ContextLogRenderer.tsx (hunk 2 above)
Comment (Rajat-Dabade) [reply frontend/src/container/LogDetailedView/ContextView/ContextLogRenderer.tsx:93]: Done ✅

File: frontend/src/container/LogDetailedView/ContextView/ContextView.styles.scss (hunk 4 above)
Comment (Rajat-Dabade) [reply frontend/src/container/LogDetailedView/ContextView/ContextView.styles.scss:4]: Removed it.

File: frontend/src/container/LogDetailedView/ContextView/ContextView.styles.scss (hunk 5 above)
Comment (Rajat-Dabade) [reply frontend/src/container/LogDetailedView/ContextView/ContextView.styles.scss:27]: Removed it. Now for height we are using `flex: 1` to take the remaining available height.

File: frontend/src/container/LogDetailedView/ContextView/ContextLogRenderer.tsx (hunk 6 above)
Comment (Rajat-Dabade) [reply frontend/src/container/LogDetailedView/ContextView/ContextLogRenderer.tsx:56]: On further investigation, I found out that we don't need to set the firstLog or the lastLog, as the dependency to fetch new data is on active-log and offset. The offset will change every time we send a new request to fetch more data. So removed it. 

https://github.com/SigNoz/signoz/assets/26198062/eeb0d2d3-096c-4101-8895-5b0647cc377e

File: frontend/src/container/LogDetailedView/ContextView/ContextLogRenderer.tsx (hunk 7 above)
Comment (Rajat-Dabade) [reply frontend/src/container/LogDetailedView/ContextView/ContextLogRenderer.tsx:63]: Removed as it was not required.

File: frontend/src/container/LogDetailedView/ContextView/ContextLogRenderer.tsx (hunk 3 above)
Comment (Rajat-Dabade) [reply frontend/src/container/LogDetailedView/ContextView/ContextLogRenderer.tsx:99]: The container wrapping the virtuoso list should be taking care of the parent height, 

Why has height is been set?
React-Virtuoso utilizes the viewport to efficiently render large lists or grids of items. It renders only the items that are currently visible within the viewport. So we are calculating the height required to render each log and setting it.

Comment (Rajat-Dabade) [reply frontend/src/container/LogDetailedView/ContextView/ContextLogRenderer.tsx:102]: Will take care of this when `endReach` and `startReach` are implemented, Then the height and number of items will be depended on the parent height. 

https://github.com/SigNoz/signoz/issues/4646

**Review by Vikrant2520 — APPROVED** [review]

**Review by YounixM — DISMISSED** [review]

**Review by YounixM — APPROVED** [review]

**Review by Vikrant2520 — APPROVED** [review]

**Pull request merged**